hack/release/prepare-changelog.sh

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env bash
+
+set -eo pipefail
+
+function echoerr {
+    >&2 echo "$@"
+}
+
+_usage="Usage: $0 --release <RELEASE> [--from-release <RELEASE>]
+Draft the changelog for a release based on PR history starting from the last release.
+For a minor release such as 1.5.0, the changelog will base on PRs merged to main branch since the fork of release-1.4.
+For a patch release such as 1.5.2, the changelog will base on PRs merged to release-1.5 branch since the release of 1.5.1.
+
+Options:
+        --release <RELEASE>        The release for which the changelog is generated.
+        --from-release <RELEASE>   The last release from which the changelog is generated. It's supposed to be used only
+                                   when running the script for a major release such as 2.0.0"
+
+function print_usage {
+  echoerr "$_usage"
+}
+
+if ! command -v gh > /dev/null; then
+  echo "Can't find 'gh' tool in PATH, please install from https://github.com/cli/cli"
+  exit 1
+fi
+
+release=""
+from_release=""
+limit=500
+
+while [[ $# -gt 0 ]]
+do
+key="$1"
+
+case $key in
+  --release)
+  release="$2"
+  shift 2
+  ;;
+  --from-release)
+  from_release="$2"
+  shift 2
+  ;;
+  -h|--help)
+  print_usage
+  exit 0
+  ;;
+  *)    # unknown option
+  echoerr "Unknown option $1"
+  exit 1
+  ;;
+esac
+done
+
+if [ "$release" == "" ]; then
+  echoerr "--release must be set"
+  exit 1
+fi
+
+major_minor_number="${release%.*}"
+major_number="${major_minor_number%.*}"
+minor_number="${major_minor_number#*.}"
+patch_number="${release##*.}"
+
+if [ "$from_release" == "" ]; then
+  if [ "$patch_number" == "0" ]; then
+    from_release="${major_number}.$(( minor_number-1 )).0"
+  else
+    from_release="${major_number}.${minor_number}.$(( patch_number-1 ))"
+  fi
+fi
+
+# For minor releases, PRs are merged to main branch.
+# For patch releases, PRs are merged to release branches.
+branch="main"
+if [ "$patch_number" != "0" ]; then
+  branch="release-${major_minor_number}"
+fi
+
+# Get the best common ancestor between the current branch and the last release.
+common_ancestor=$(git merge-base "v${from_release}" "${branch}")
+# Get the merge time of the common ancestor and use it as the start time of the current release.
+release_start_time=$(gh pr list --state merged --search "${common_ancestor}" --json mergedAt -q .[0].mergedAt)
+
+# Do not generate the title for patch releases.
+if [ "$patch_number" == "0" ]; then
+  echo "# Changelog ${major_minor_number}"
+  echo ""
+fi
+echo "## ${release} - $(date +%F)"
+echo ""
+echo "### Added"
+echo ""
+
+echo "### Changed"
+echo ""
+# Put the changes under "Changed" first, release manager needs to move them to appropriate sections manually.
```

Comment:
we could adopt a model which is closer to what K8s does, it would make the changelog generation easier
It seems they have the following "sections": Deprecation, API Change, Feature, Documentation, Failing Test, Bug or Regression, Other. I believe these map to Github labels directly.

we can also look into doing this at a future date

Reply:
Yes, it's a good idea. Will use separate PRs to improve it.